File: PermutationInString.java
import java.util.Arrays;
import java.util.stream.IntStream;

public class PermutationInString {
    public boolean checkInclusion(String s1, String s2) {

        if(s2.length() < s1.length()){
            return false;
        }
        int[] s1FreqMap = new int[26];
        int[] s2FreqMap = new int[26];

        IntStream.range(0, s1.length())
            .forEach(index -> {
                s1FreqMap[s1.charAt(index) - 'a']++;
                s2FreqMap[s2.charAt(index) - 'a']++;
            });
        
        if(Arrays.equals(s1FreqMap, s2FreqMap)){
            return true;
        }

        for(int i=0, j=s1.length(); j<s2.length(); j++,i++){
            s2FreqMap[s2.charAt(i)-'a']--;            
            s2FreqMap[s2.charAt(j)-'a']++;

            if(Arrays.equals(s1FreqMap, s2FreqMap)){
                return true;
            }
        }

        return false;
        
    }

    public static void main(String[] args) {
        String s1 = "ab", s2 = "eidbaooo";

        PermutationInString permutationInString = new PermutationInString();
        boolean s = permutationInString.checkInclusion(s1, s2);
        System.out.println(s);
    }
}

File: permutationInString.go
package main

import "fmt"

func checkInclusion(s1 string, s2 string) bool {
	var s1FreqMap [26]int32
	var s2FreqMap [26]int32

	if len(s1) > len(s2) {
		return false
	}

	s1Rune := []rune(s1)
	s2Rune := []rune(s2)

	for i := 0; i < len(s1Rune); i++ {
		s1FreqMap[s1Rune[i]-'a']++
		s2FreqMap[s2Rune[i]-'a']++
	}

	if s1FreqMap == s2FreqMap {
		return true
	}

	for i, j := 0, len(s1); j < len(s2); i, j = i+1, j+1 {

		s2FreqMap[s2Rune[i]-'a']--
		s2FreqMap[s2Rune[j]-'a']++
		if s1FreqMap == s2FreqMap {
			return true
		}
	}

	return false

}

func main() {
	s1 := "ab"
	s2 := "eidbaooo"

	fmt.Print(checkInclusion(s1, s2))
}
